# In-Class Coding Lab: Data Analysis with Pandas

In this lab, we will perform a data analysis on the **RMS Titanic** passenger list. The RMS Titanic is one of the most famous ocean liners in history. On April 15, 1912 it sank after colliding with an iceberg in the North Atlantic Ocean. To learn more, read here: https://en.wikipedia.org/wiki/RMS_Titanic 

Our goal today is to perform a data analysis on a subset of the passenger list. We're looking for insights as to which types of passengers did and didn't survive. Women? Children? 1st Class Passengers? 3rd class? Etc. 

I'm sure you've heard the expression often said during emergencies: "Women and Children first" Let's explore this data set and find out if that's true!

Before we begin you should read up on what each of the columns mean in the data dictionary. You can find this information on this page: https://www.kaggle.com/c/titanic/data 


## Loading the data set

First we load the dataset into a Pandas `DataFrame` variable. The `sample(10)` method takes a random sample of 10 passengers from the data set.

In [2]:
import pandas as pd
import numpy as np

# this turns off warning messages
import warnings
warnings.filterwarnings('ignore')

passengers = pd.read_csv('CCL-titanic.csv')
passengers.sample(10)

PassengerId  Survived  Pclass  \
656          657         0       3   
527          528         0       1   
774          775         1       2   
387          388         1       2   
560          561         0       3   
886          887         0       2   
571          572         1       1   
80            81         0       3   
49            50         0       3   
725          726         0       3   

                                              Name     Sex   Age  SibSp  \
656                          Radeff, Mr. Alexander    male   NaN      0   
527                             Farthing, Mr. John    male   NaN      0   
774          Hocking, Mrs. Elizabeth (Eliza Needs)  female  54.0      1   
387                               Buss, Miss. Kate  female  36.0      0   
560                       Morrow, Mr. Thomas Rowan    male   NaN      0   
886                          Montvila, Rev. Juozas    male  27.0      0   
571  Appleton, Mrs. Edward Dale (Charlotte Lamson)  female  53.0      2   
80                            Waelens, Mr. Achille    male  22.0      0   
49   Arnold-Franchi, Mrs. Josef (Josefine Franchi)  female  18.0      1   
725                            Oreskovic, Mr. Luka    male  20.0      0   

     Parch    Ticket      Fare Cabin Embarked  
656      0    349223    7.8958   NaN        S  
527      0  PC 17483  221.7792   C95        S  
774      3     29105   23.0000   NaN        S  
387      0     27849   13.0000   NaN        S  
560      0    372622    7.7500   NaN        Q  
886      0    211536   13.0000   NaN        S  
571      0     11769   51.4792  C101        S  
80       0    345767    9.0000   NaN        S  
49       0    349237   17.8000   NaN        S  
725      0    315094    8.6625   NaN        S

## How many survived?

One of the first things we should do is figure out how many of the passengers in this data set survived. Let's start with isolating just the `'Survivied'` column into a series:

In [3]:
passengers['Survived'].sample(10)

165    1
38     0
411    0
632    1
890    0
288    1
475    0
685    0
505    0
461    0
Name: Survived, dtype: int64

There's too many to display so we just display a random sample of 10 passengers. 

- 1 means the passenger survivied
- 0 means the passenger died

What we really want is to count the number of survivors and deaths. We do this by querying the `value_counts()` of the `['Survived']` column, which returns a `Series` of counts, like this:

In [4]:
passengers['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

Only 342 passengers survived, and 549 perished. Let's observe this same data as percentages of the whole. We do this by adding the `normalize=True` named argument to the `value_counts()` method.

In [5]:
passengers['Survived'].value_counts(normalize=True)

0    0.616162
1    0.383838
Name: Survived, dtype: float64

**Just 38% of passengers in this dataset survived.**

### Now you Try it!

**FIRST** Write a Pandas expression to display counts of males and female passengers using the `Sex` variable:

In [8]:
# todo write code here
passengers['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

**NEXT** Write a Pandas expression to display male /female passenger counts as a percentage of the whole number of passengers in the data set.

In [9]:
# todo write code here
passengers['Sex'].value_counts(normalize=True)

male      0.647587
female    0.352413
Name: Sex, dtype: float64

If you got things working, you now know that **35% of passengers were female**.

## Who survivies? Men or Women?

We now know that 35% of the passengers were female, and 65% we male. 

**The next think to think about is how do survivial rates affect these numbers? **

If the ratio is about the same for surviviors only, then we can conclude that your **Sex** did not play a role in your survival on the RMS Titanic. 

Let's find out.

In [10]:
survivors = passengers[passengers['Survived'] ==1]
survivors['PassengerId'].count()

342

Still **342** like we discovered originally. Now let's check the **Sex** split among survivors only:

In [11]:
survivors['Sex'].value_counts()

female    233
male      109
Name: Sex, dtype: int64

WOW! That is a huge difference! But you probably can't see it easily. Let's represent it in a `DataFrame`, so that it's easier to visualize:

In [12]:
sex_all_series = passengers['Sex'].value_counts()
sex_survivor_series = survivors['Sex'].value_counts()

sex_comparision_df = pd.DataFrame({ 'AllPassengers' : sex_all_series, 'Survivors' : sex_survivor_series })
sex_comparision_df['SexSurvivialRate'] = sex_comparision_df['Survivors'] / sex_comparision_df['AllPassengers']
sex_comparision_df

AllPassengers  Survivors  SexSurvivialRate
female            314        233          0.742038
male              577        109          0.188908

 **So, females had a 74% survival rate. Much better than the overall rate of 38%**
 
We should probably briefly explain the code above. 

- The first two lines get a series count of all passengers by Sex (male / female) and count of survivors by sex
- The third line creates DataFrame. Recall a pandas dataframe is just a dict of series. We have two keys 'AllPassengers' and 'Survivors'
- The  fourth line creates a new column in the dataframe which is just the survivors / all passengers to get the rate of survival for that Sex.

## Feature Engineering: Adults and Children

Sometimes the variable we want to analyze is not readily available, but can be created from existing data. This is commonly referred to as **feature engineering**. The name comes from machine learning where we use data called *features* to predict an outcome. 

Let's create a new feature called `'AgeCat'` as follows:

- When **Age** <=18 then 'Child'
- When **Age** >18 then 'Adult'

This is easy to do in pandas. First we create the column and set all values to `np.nan` which means 'Not a number'. This is Pandas way of saying no value. Then we set the values based on the rules we set for the feature.

In [28]:
passengers['AgeCat'] = np.nan # Not a number
passengers['AgeCat'][ passengers['Age'] <=18 ] = 'Child'
passengers['AgeCat'][ passengers['Age'] > 18 ] = 'Adult'
passengers.sample(5)

PassengerId  Survived  Pclass                           Name     Sex  \
267          268         1       3       Persson, Mr. Ernst Ulrik    male   
318          319         1       1       Wick, Miss. Mary Natalie  female   
843          844         0       3     Lemberopolous, Mr. Peter L    male   
350          351         0       3         Odahl, Mr. Nils Martin    male   
24            25         0       3  Palsson, Miss. Torborg Danira  female   

      Age  SibSp  Parch  Ticket      Fare Cabin Embarked AgeCat  
267  25.0      1      0  347083    7.7750   NaN        S  Adult  
318  31.0      0      2   36928  164.8667    C7        S  Adult  
843  34.5      0      0    2683    6.4375   NaN        C  Adult  
350  23.0      0      0    7267    9.2250   NaN        S  Adult  
24    8.0      3      1  349909   21.0750   NaN        S  Child

Let's get the count and distrubutions of Adults and Children on the passenger list.

In [29]:
passengers['AgeCat'].value_counts()

Adult    575
Child    139
Name: AgeCat, dtype: int64

And here's the percentage as a whole:

In [30]:
passengers['AgeCat'].value_counts(normalize=True)

Adult    0.805322
Child    0.194678
Name: AgeCat, dtype: float64

So close to **80%** of the passengers were adults. Once again let's look at the ratio of `AgeCat` for survivors only. If your age has no bearing of survivial, then the rates should be the same. 

Here's the counts of Adult / Children among the survivors only:

In [31]:
survivors = passengers[passengers['Survived'] ==1]
survivors['AgeCat'].value_counts()

Adult    220
Child     70
Name: AgeCat, dtype: int64

### Now You Try it!

Calculate the `AgeCat` survival rate, similar to how we did for the `SexSurvivalRate`. 

In [34]:
agecat_all_series = passengers['AgeCat'].value_counts()
agecat_survivor_series = survivors['AgeCat'].value_counts()

# todo make a data frame, add AgeCatSurvivialRate column, display dataframe 
agecat_df = pd.DataFrame({"All" : agecat_all_series, "Survivors" : agecat_survivor_series})
agecat_df['Age Survival Rate'] = agecat_df['Survivors'] / agecat_df ['All']
agecat_df

All  Survivors  Age Survival Rate
Adult  575        220           0.382609
Child  139         70           0.503597

**So, children had a 50% survival rate, better than the overall rate of 38%**

## So, women and children first?

It looks like the RMS really did have the motto: "Women and Children First."

Here's our insights. We know:

- If you were a passenger, you had a 38% chance of survival.
- If you were a female passenger, you had a 74% chance of survival.
- If you were a child passenger, you had a 50% chance of survival. 


### Now you try it for Passenger Class

Repeat this process for `Pclass` The passenger class variable. Display the survival rates for each passenger class. What does the information tell you about passenger class and survival rates?

I'll give you a hint... "Money Talks"


In [35]:
# todo: repeat the analysis in the previous cell for Pclass 
pclass_series = passengers['Pclass'].value_counts()
pclass_survivor_series = survivors['Pclass'].value_counts()

pclass_df = pd.DataFrame({'All' : pclass_series, 'Survivors' : pclass_survivor_series})
pclass_df['Class Survival Rate'] = pclass_df['Survivors'] / pclass_df['All']
pclass_df

All  Survivors  Class Survival Rate
1  216        136             0.629630
2  184         87             0.472826
3  491        119             0.242363

https://youtu.be/dqB-EMqpsUA